src/client_side_request.cc

```diff
@@ -1794,9 +1794,14 @@ ClientHttpRequest::doCallouts()
             ACLFilledChecklist ch(NULL, request, NULL);
             ch.src_addr = request->client_addr;
             ch.my_addr = request->my_addr;
-            nfmark_t mark = aclMapNfmark(Ip::Qos::TheConfig.nfmarkToClient, &ch);
-            if (mark)
+            acl_nfmark *l = aclMapNfmark(Ip::Qos::TheConfig.nfmarkToClient, &ch);
+            if (l) {
+                // Clear the bits given by the mask in existing CONNMARK
+                // and XOR it with a new mark.
+                nfmark_t mark = getConn()->clientConnection->nfmark & ~(l->nfmask);
+                mark ^= l->nfmark;
```

Comment:
I have two concerns here:

1. To an non-expert like me, it seems that the configured mask should apply to the configured `nfmark` value as well.
2. If `nfmask` is applied to `nfmark`, then the XOR part seems unnecessary complex to me. The `~nfmask` clearing above guarantees that all the relevant bits in `mark` are going to be zero. Similarly, all the irrelevant bits in `nfmark` will be zero due to (1). Thus, applying a simpler OR logic is sufficient: `0 xor x == 0 or x`

In summary, I wonder whether you want an OR operation with a mask (instead of XOR without a mask):

```
mark |= (l->nfmark & l->nfmask);
```

This is not my area of expertise so if XOR-without-a-mask is how we are supposed to implement this to accommodate some unusual/advanced use cases, please disregard the above concern, with my apologies. Otherwise, please also adjust the documentation to match both changes discussed above.

Reply:
I've borrowed the logic from iptables' 'set-xmark' option that works as described here: https://stackoverflow.com/a/22003527

Probably the documentation should be adjusted to cover all possible cases.

Reply:
Aha, I think I know where the core of the problem is. We need to make a choice between two very different marking semantics. You picked complex/adjustment-focused `--set-xmark` semantics over the simple/assignment-focused `--set-mark` semantics. If you look at the actual [documentation](http://ipset.netfilter.org/iptables-extensions.man.html) that the Stackflow article is quoting, you will see both operations defined:

```
--set-xmark value[/mask]
    Zero out the bits given by mask and XOR value into the ctmark. 

The following mnemonics are available for --set-xmark:

--and-mark bits
    Binary AND the ctmark with bits. (Mnemonic for --set-xmark 0/invbits, where invbits is the binary negation of bits.) 
--or-mark bits
    Binary OR the ctmark with bits. (Mnemonic for --set-xmark bits/bits.) 
--xor-mark bits
    Binary XOR the ctmark with bits. (Mnemonic for --set-xmark bits/0.) 
--set-mark value[/mask]
    Set the connection mark. If a mask is specified then only those bits set in the mask are modified. 
```

As I expected, the set-xmark semantics is very powerful -- it can do a lot of fancy operations. However, AFAICT, it cannot fully implement the set-mark semantics! Notice how the mnemonics section of the above documentation gives equivalent expressions for all other options _except_ `--set-mark`? The iptables documentation misleads -- the set-mark option is not a mnemonic for a subset of set-xmark use cases. It is a stand-alone option!

Now that we understand that there are two different operations, we have to pick the one we should implement in Squid. Before this PR, Squid supported both operations because when the mask is all-ones, there is no difference between the two. However, there is a difference going forward when the masks become non-trivial.

* If we implement just the set-xmark semantics, then the admin will not be able to tell Squid to set the last byte of the mark to `0xB` without touching anything else.
* If we implement just the set-mark semantics, then the admin will not be able to tell Squid, for example, to add  `0xB` bits to the existing mask bits without turning any existing bits off. There are other mask adjustment operations that the admin will not be able to do either.

The Squid directive is _not_ named `clientside_xmark` -- there is no visual warning for the admin that something complex is going on.

Given all of the above considerations/concerns, we have two reasonable choices IMHO:

1. Implement just the set-mark algorithm. This option covers only the simple marking operations. It is easy to adjust your PR to implement this option.
2. Implement both set-xmark and set-mark algorithms with some syntax sugar (or a new directive) to distinguish the two. This option covers both complex and simple marking operations. Its implementation requires significant additional work.

If _you_ need the set-xmark power, then I suggest moving towards option 2. Otherwise, we can start with option 1 for now and add set-xmark support later when somebody needs it. This is just my non-expert opinion. Others may suggest a different decision-making algorithm (and different implementation options).

What do _you_ think? And do you want to wait for other reviews before modifying the PR further?

Reply:
Yes, it seems that I've tried to do it the hard way. Currently I think that set-mark logic would be a better and more obvious option.

Reply:
I'll try to fix current issues (and change the logic, of course) and then we can get back to review.